lib/spack/spack/cmd/testinstall.py

```diff
@@ -0,0 +1,181 @@
+
+def setup_parser(subparser):
+    #subparser.add_argument(
+    #    '-i', '--ignore-dependencies', action='store_true', dest='ignore_deps',
+    #    help="Do not try to install dependencies of requested packages.")
+    
+    subparser.add_argument(
+        '-j', '--jobs', action='store', type=int,
+        help="Explicitly set number of make jobs.  Default is #cpus.")
+        
+    #always false for test
+    #subparser.add_argument(
+    #    '--keep-prefix', action='store_true', dest='keep_prefix',
+    #    help="Don't remove the install prefix if installation fails.")
+    
+    #always true for test
+    #subparser.add_argument(
+    #    '--keep-stage', action='store_true', dest='keep_stage',
+    #    help="Don't remove the build stage if installation succeeds.")
+    
+    subparser.add_argument(
+        '-n', '--no-checksum', action='store_true', dest='no_checksum',
+        help="Do not check packages against checksum")
+    subparser.add_argument(
+        '-v', '--verbose', action='store_true', dest='verbose',
+        help="Display verbose build output while installing.")
+    
+    #subparser.add_argument(
+    #    '--fake', action='store_true', dest='fake',
+    #    help="Fake install.  Just remove the prefix and touch a fake file in it.")
+    
+    subparser.add_argument(
+        'output', help="test output goes in this file")
+    
+    subparser.add_argument(
+        'packages', nargs=argparse.REMAINDER, help="specs of packages to install")
+
+
+class JunitResultFormat(object):
+    def __init__(self):
+        self.root = ET.Element('testsuite')
+        self.tests = []
+        
+    def addTest(self, buildId, passed=True, buildInfo=None):
+        self.tests.append((buildId, passed, buildInfo))
+    
+    def writeTo(self, stream):
+        self.root.set('tests', '{0}'.format(len(self.tests)))
+        for buildId, passed, buildInfo in self.tests:
+            testcase = ET.SubElement(self.root, 'testcase')
+            testcase.set('classname', buildId.name)
+            testcase.set('name', buildId.stringId())
+            if not passed:
+                failure = ET.SubElement(testcase, 'failure')
+                failure.set('type', "Build Error")
+                failure.text = buildInfo
+        ET.ElementTree(self.root).write(stream)
+
+
+class BuildId(object):
+    def __init__(self, name, version, hashId):
+        self.name = name
+        self.version = version
+        self.hashId = hashId
+    
+    def stringId(self):
+        return "-".join(str(x) for x in (self.name, self.version, self.hashId))
+
+
+def createTestOutput(spec, handled, output):
+    if spec in handled:
+        return handled[spec]
+    
+    childSuccesses = list(createTestOutput(dep, handled, output) 
+            for dep in spec.dependencies.itervalues())
+    package = spack.db.get(spec)
+    handled[spec] = package.installed
+    
+    if all(childSuccesses):
+        bId = BuildId(spec.name, spec.version, spec.dag_hash())
```

Comment:
I think this build id makes sense for now.  It would be good to eventually allow this to be passed in as a parameter.  The user could supply a format string indicating how they want the build id to look for a particular test.  Let's wait on that because it depends on some support Matt has added in another PR, but keep it in mind.